src/parse.c

```diff
@@ -1105,12 +1105,76 @@ handle_addresses(yaml_document_t* doc, yaml_node_t* node, const void* _, GError*
     return handle_generic_addresses(doc, node, TRUE, &(cur_netdef->ip4_addresses), &(cur_netdef->ip6_addresses), error);
 }
 
+const static char *GW_MULTIPLE_IP_ERROR = "gateway: there must be at most one of each IP type";
+
+struct gw_parse_result {
+	const char *gw4;
+	const char *gw6;
+};
+
+static gboolean
+handle_single_gateway(yaml_node_t* node, GError** error, struct gw_parse_result *out)
+{
+    const char *addr;
+
+    assert_type(node, YAML_SCALAR_NODE);
+
+    addr = scalar(node);
+    if (is_ip4_address(addr)) {
+	    if (out->gw4)
+		    return yaml_error(node, error, GW_MULTIPLE_IP_ERROR);
+	    out->gw4 = addr;
+    } else if (is_ip6_address(addr)) {
+	    if (out->gw6)
+		    return yaml_error(node, error, GW_MULTIPLE_IP_ERROR);
+	    out->gw6 = addr;
+    } else
+	    return yaml_error(node, error, "invalid IP address '%s'", addr);
+    return TRUE;
+}
+
+static gboolean
+handle_gateway(yaml_document_t* doc, yaml_node_t* node, const void* _, GError** error)
+{
+    struct gw_parse_result res = {NULL, NULL};
+
+    if (node->type == YAML_SCALAR_NODE) {
+        if (!handle_single_gateway(node, error, &res))
+            return FALSE;
+    } else {
+        yaml_node_item_t *iter;
+        assert_type(node, YAML_SEQUENCE_NODE);
+
+        for (iter = node->data.sequence.items.start; iter < node->data.sequence.items.top; iter++) {
+            if (!handle_single_gateway(yaml_document_get_node(doc, *iter), error, &res))
+                return FALSE;
+        }
+    }
+
+    if (res.gw4) {
+        if (cur_netdef->gateway4)
+            g_assert_cmpstr(cur_netdef->gateway4, ==, res.gw4);
+        else
+            cur_netdef->gateway4 = g_strdup(res.gw4);
+    }
+
+    if (res.gw6) {
```

Comment:
I wonder if we need the `res.gw4/6` helper variable at all... I see how it nicely avoids some multi-pass parsing issues, as res is initialized as `{NULL, NULL}` for each new parsing pass. But then down here we still need to check if it was assigned already (to an equal value) in a previous parsing pass.

I think we should be able to write directly into `cur_netdef->gateway4/6` from inside `handle_single_gateway()`, comparing if the value already exists and is equal (=> skip). I guess this would not throw an error for cases like this: `gateway: [1.2.3.4, 1.2.3.4]` as this would look like being a 2nd parsing pass. But this is obviously wrong in the YAML and makes no difference, really (setting the same gateway twice). I wonder if we should make this trade-off in order to save passing around the extra `_gw_parse_result` struct. What do you think?

Reply:
I used this approach initially, although I didn't think of strcmp'ing directly to the gateway4/6, which of course meant that the wireguard tests were failing.

However, in retrospect, I still prefer using the _gw_parse_result local structure to store the intermediate data while we're still validating the content of the field. This makes the failure case much easier to reason about. As a bonus, it catches an obvious configuration mistake which, while harmless in itself might prompt the user to look at their code a bit more carefully ;)

Reply:
ACK. Properly handling the (obvious) error case is a valid reason to keep that struct around, IMO.